### PR TITLE
rustc: Load the `rustc_trans` crate at runtime

### PR DESCRIPTION
**Pull request by alexcrichton**
Building on the work of #45684 this commit updates the compiler to
unconditionally load the `rustc_trans` crate at runtime instead of linking to it
at compile time. The end goal of this work is to implement #46819 where rustc
will have multiple backends available to it to load.

This commit starts off by removing the `extern crate rustc_trans` from the
driver. This involved moving some miscellaneous functionality into the
`TransCrate` trait and also required an implementation of how to locate and load
the trans backend. This ended up being a little tricky because the sysroot isn't
always the right location (for example `--sysroot` arguments) so some extra code
was added as well to probe a directory relative to the current dll (the
rustc_driver dll).

Rustbuild has been updated accordingly as well to have a separate compilation
invocation for the `rustc_trans` crate and assembly it accordingly into the
sysroot. Finally, the distribution logic for the `rustc` package was also
updated to slurp up the trans backends folder.

A number of assorted fallout changes were included here as well to ensure tests
pass and such, and they should all be commented inline.

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @Mark-Simulacrum 

cc @michaelwoerister, @eddyb, @bjorn3

**Comment by alexcrichton** [comment]
The next plans for https://github.com/rust-lang/rust/issues/46819 are to add rustbuild configuration and support (as well as another submodule) for Emscripten's backend, integrate that for Emscripten, and then we should be good to upgrade to LLVM 5!

File: src/librustc_driver/lib.rs [hunk 1]
```diff
@@ -176,57 +179,242 @@ pub fn run<F>(run_compiler: F) -> isize
...
             }
         }
-        Err(err) => {
-            sess.fatal(&format!("Couldnt load codegen backend {:?}: {:?}", backend_name, err));
+    }
+
+    let sysroot = sysroot_candidates.iter()
+        .map(|sysroot| {
+            let libdir = filesearch::relative_target_lib_path(&sysroot, &target);
+            sysroot.join(&libdir).join("trans-backends")
```
Comment (eddyb) [inline src/librustc_driver/lib.rs]: Just a nitpick, but could you use "codegen backend" instead of "trans backend" for everything external to the compiler itself? It can be changed later as part of #45274, so it doesn't really matter as long as we don't stabilize anything based on it.

File: src/librustc_trans/lib.rs [hunk 2]
```diff
@@ -149,13 +149,16 @@ impl !Send for LlvmTransCrate {} // Llvm is on a per-thread basis
...
-    pub fn new(sess: &Session) -> Box<TransCrate> {
-        llvm_util::init(sess); // Make sure llvm is inited
+    pub fn new() -> Box<TransCrate> {
         box LlvmTransCrate(())
     }
 }
 
 impl TransCrate for LlvmTransCrate {
+    fn init(&self, sess: &Session) {
+        llvm_util::init(sess); // Make sure llvm is inited
```
Comment (eddyb) [inline src/librustc_trans/lib.rs]: I think this is fine for now but in the future we might want to have a `Session` created by the time we create this trait object (we'll need at least the target spec loaded).

Comment (bjorn3) [reply src/librustc_trans/lib.rs]: Maybe let `__rustc_codegen_backend` take a `Option<&Session>` as argument and pass None when we are only using it for printing version info and stuff like that.

Comment (eddyb) [reply src/librustc_trans/lib.rs]: No, I think we can always create a `Session`, there shouldn't really be a significant cost to it since we're already doing the argument parsing and we have to load the target specs anyway.

Comment (alexcrichton) [reply src/librustc_trans/lib.rs]: Ah yeah so the two locations this was necessary (not taking a `Session` on construction) was when we print the version and when we print the LLVM passes. There's also one where we need to construct a list of all error codes but that's commented out for now...

I it'd be possible to have a `Session` here, but it'd be a "dummy session" in some cases because I don't think we want to generate an error for failure to parse command line arguments before we print the version. Either way I'd be fine though!

Comment (eddyb) [reply src/librustc_trans/lib.rs]: We don't fully parse arguments before processing the version? Is it a separate codepath?

Comment (alexcrichton) [reply src/librustc_trans/lib.rs]: @eddyb yeah at least for now the codepaths are pretty separate, I think it's *while* we're parsing arguments we'd delegate out to this.

It does mean that various print modes won't work with custom backends immediately but that's probably ok for now. Eventually we may be able to just require a couple of symbols to be defined and continue to require a `Session` for this.

Or a different trait object is returned perhaps!

Comment (eddyb) [reply src/librustc_trans/lib.rs]: I don't know, it seems fragile - are you saying `rustc --target=...emscripten -vV` and `rustc -vV --target=...emscripten` would behave differently? We could certainly *defer* CLI errors without preventing the creation of a `Session`, and not print them in certain situations.

Comment (alexcrichton) [reply src/librustc_trans/lib.rs]: Oh no both of those wouldn't use `--target` at all, the `-vV` is detected early. It does mean though that both of those would not print the version of the emscripten backend (or print passes from it).

Do you think this should block this landing? I figured this was esoteric enough that there's no need to fix so soon.

Comment (eddyb) [reply src/librustc_trans/lib.rs]: Well this PR doesn't use target specs yet, so it's not necessary *for now*. But I am curious, because I was hoping for a straight-forward implementation, is the "printing the version skips checking the rest of the CLI arguments" guaranteed/relied upon?

Comment (alexcrichton) [reply src/librustc_trans/lib.rs]: I think it basically would just require more refactoring. We'd have to, for example, initialize the `Session` in stages, first with a bunch of known options, then do some processing, then aftewards actually validate everything. AFAIK it's not really set up to do that yet.

Comment (eddyb) [reply src/librustc_trans/lib.rs]: Can't we validate everything early and buffer the errors or something?

Comment (alexcrichton) [reply src/librustc_trans/lib.rs]: I guess? Again though I didn't get the impression that the argument parsing was at all ready for this sort of refactoring, in the sense that it seemed outside the scope of this PR

Comment (eddyb) [reply src/librustc_trans/lib.rs]: Yeah, I don't want to block *this* PR, this is more for future reference.

File: src/librustc_driver/lib.rs [hunk 3]
```diff
@@ -1316,8 +1504,8 @@ pub fn diagnostics_registry() -> errors::registry::Registry {
     all_errors.extend_from_slice(&rustc_typeck::DIAGNOSTICS);
     all_errors.extend_from_slice(&rustc_resolve::DIAGNOSTICS);
     all_errors.extend_from_slice(&rustc_privacy::DIAGNOSTICS);
-    #[cfg(feature="llvm")]
-    all_errors.extend_from_slice(&rustc_trans::DIAGNOSTICS);
+    // FIXME: need to figure out a way to get these back in here
+    // all_errors.extend_from_slice(get_trans(sess).diagnostics());
```
Comment (Mark-Simulacrum) [inline src/librustc_driver/lib.rs]: This might be a function that should lie on the trait -- it seems like it should anyway, or maybe the trait could return a vec of errors that need to be added.

File: src/test/run-pass-fulldeps/create-dir-all-bare.rs [hunk 4]
```diff
@@ -8,18 +8,12 @@
...
 use std::env;
 use std::fs;
-use tempdir::TempDir;
+use std::path::PathBuf;
 
 fn main() {
-    let td = TempDir::new("create-dir-all-bare").unwrap();
-    env::set_current_dir(td.path()).unwrap();
+    let path = PathBuf::from(env::var_os("RUST_TEST_TMPDIR").unwrap());
+    env::set_current_dir(&path).unwrap();
```
Comment (Mark-Simulacrum) [inline src/test/run-pass-fulldeps/create-dir-all-bare.rs]: Why are these changes necessary? Can we no longer link to tempdir from tests? Or is it that tempdir wasn't being rebuilt because rustbuild is buggy (I would *not* be surprised).

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]
Some general nits. Overall approach looks okay to me, though I'm not super happy about having to add another step to compile.rs... seems okay though.

File: src/librustc_driver/lib.rs [hunk 5]
```diff
@@ -176,57 +179,242 @@ pub fn run<F>(run_compiler: F) -> isize
...
+    // For now we only allow this function to be called once as it'll dlopen a
+    // few things, which seems to work best if we only do that once. In
+    // general this assertion never trips due to the once guard in `get_trans`,
+    // but there's a few manual calls to this function in this file we protect
+    // against.
+    static LOADED: AtomicBool = ATOMIC_BOOL_INIT;
+    assert!(!LOADED.swap(true, Ordering::SeqCst),
+            "cannot load the default trans backend twice");
+
+    if cfg!(test) {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Certainly! I've added a comment

File: src/librustc_driver/lib.rs [hunk 6]
```diff
@@ -176,57 +179,242 @@ pub fn run<F>(run_compiler: F) -> isize
...
+
+    let mut file: Option<PathBuf> = None;
+
+    for entry in d.filter_map(|e| e.ok()) {
+        let path = entry.path();
+        let filename = match path.file_name().and_then(|s| s.to_str()) {
+            Some(s) => s,
+            None => continue,
+        };
+        if !filename.starts_with(DLL_PREFIX) || !filename.ends_with(DLL_SUFFIX) {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Hm yeah that's the condition I'm going for, but I think that's what's here as well? I'm looking for

```
if !(correct_prefix && correct_suffix) {
    continue
}
```

which should be the same as 

```
if !correct_prefix || !correct_suffix {
   continue
}
```

File: src/librustc_driver/lib.rs [hunk 7]
```diff
@@ -176,57 +179,242 @@ pub fn run<F>(run_compiler: F) -> isize
...
+    #[cfg(unix)]
+    fn current_dll_path() -> Option<PathBuf> {
+        use std::ffi::{OsStr, CStr};
+        use std::os::unix::prelude::*;
+
+        unsafe {
+            let addr = current_dll_path as usize as *mut _;
+            let mut info = mem::zeroed();
+            if libc::dladdr(addr, &mut info) == 0 {
+                info!("dladdr failed: {}", io::Error::last_os_error());
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Turns out, no!

File: src/librustc_driver/lib.rs [hunk 8]
```diff
@@ -176,57 +179,242 @@ pub fn run<F>(run_compiler: F) -> isize
...
+                return None
+            }
+            if info.dli_fname.is_null() {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: I think that's ok for us? I haven't used this sort of dynamic loading much before so I suspect that we'll probably find a bug or two along the way, but I think this specific situation may fall under the category of "turns out ok given our usage" maybe?

File: src/librustc_driver/lib.rs (hunk 3 above)
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Heh it actually is! Unfortunately though there's not a great way to load it here, so I figured that for the one extended error in librustc_trans we could fudge it for now and figure it out later.

File: src/test/run-pass-fulldeps/create-dir-all-bare.rs (hunk 4 above)
Comment (alexcrichton) [reply src/test/run-pass-fulldeps/create-dir-all-bare.rs]: The dependencies of librustc_trans are no longer available in the sysroot (as we just manually copy the one DLL), and one of its dependencies was `tempdir`. (no other crates in rustc use `tempdir`)

**Comment by alexcrichton** [comment]
Updated!

File: src/librustc_driver/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/librustc_driver/lib.rs]: Hm, yeah. Maybe clearer to write it that way though? Probably a minor point and not entirely relevant anyway. I was confused; but I don't think there's anyway to make this nonconfusing really.

File: src/librustc_driver/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/librustc_driver/lib.rs]: So I guess what I'm asking is why do we check `dli_fname` and not `dli_sname`? Judging by the documentation, what we check today isn't going to determine whether the symbol was found or not, which I believe we do want to determine.

Agreed though that I'm not too worried about hashing this out now, if it works in your usage then seems good enough.

File: src/librustc_driver/lib.rs (hunk 3 above)
Comment (Mark-Simulacrum) [reply src/librustc_driver/lib.rs]: Could we just load all of the diagnostics, independent of backend, and then just some of them would be unused?

File: src/test/run-pass-fulldeps/create-dir-all-bare.rs (hunk 4 above)
Comment (Mark-Simulacrum) [reply src/test/run-pass-fulldeps/create-dir-all-bare.rs]: As I recall, the previous solution to this "problem" would be to move tempdir to a dependency of `librustc`. I'm fine with this too though.

**Review by Mark-Simulacrum — APPROVED** [review]
Changes seem good other than the comments I responded to.

File: src/librustc_driver/lib.rs (hunk 3 above)
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Unfortunately that's also gonna be difficult I think. I believe that with this PR as-is we could do that, but in the future I think we'll only want to dlopen one version of LLVM into the current process, and at this point in time we don't actually know the trans backend we'd otherwise be using. That means that with the Emscripten target we'd load the normal LLVM backend to load diagnostics and then load the Emscripten backend to actually do trans. I *think* that it would work ok? The "global" nature of the dlopen though required here though may throw a wrench into that..

In essence though I don't think there's a quick fix or an easy workaround for this, we'll want to deal with it later I think.

File: src/test/run-pass-fulldeps/create-dir-all-bare.rs (hunk 4 above)
Comment (alexcrichton) [reply src/test/run-pass-fulldeps/create-dir-all-bare.rs]: It's true yeah we could do that but I'm also always a fan of tests with fewer dependencies :)

File: src/librustc_driver/lib.rs (hunk 3 above)
Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Oh also note that I added a dynamic assertion for this, the business with the `LOADED` constant, which asserts that we only load a backend once.

Comment (Mark-Simulacrum) [reply src/librustc_driver/lib.rs]: Sure, yeah, I understand that we don't want to dlopen multiple rustc_trans backends, but I was rather suggesting that we put all the diagnostic definition in `driver` and then the backend would use whichever ones it needed, though not all. This does increase the API surface of driver.

Comment (eddyb) [reply src/librustc_driver/lib.rs]: The diagnostics can probably be moved to the `librustc_mir/monomorphize` instance collector, so they're shared by all backends.

Comment (alexcrichton) [reply src/librustc_driver/lib.rs]: Ah yes indeed, such a strategy would for sure work!

File: src/librustc_driver/lib.rs [hunk 9]
```diff
@@ -176,57 +179,247 @@ pub fn run<F>(run_compiler: F) -> isize
...
+    };
+    unsafe {
+        match lib.symbol("__rustc_codegen_backend") {
+            Ok(f) => {
+                mem::forget(lib);
+                mem::transmute::<*mut u8, _>(f)
+            }
+            Err(e) => {
+                let err = format!("couldn't load codegen backend as it \
+                                   doesn't export the `__rustc_backend_new` \
```
Comment (pepyakin) [inline src/librustc_driver/lib.rs]: `__rustc_codegen_backend` symbol?

**Comment by kennytm** [comment]
`dist-i686-linux` linker failures when cross-compiling `stage1-rustc_trans` from 64-bit to 32-bit.

<details>

```
[01:38:24]    Compiling rustc_llvm v0.0.0 (file:///checkout/src/librustc_llvm)
[01:41:32] error: linking with `cc` failed: exit code: 1
[01:41:32]   |
<snip>
[01:41:32]   = note: /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../x86_64-unknown-linux-gnu/bin/ld: skipping incompatible /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../lib64/libgcc_s.so when searching for -lgcc_s
[01:41:32]           /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../x86_64-unknown-linux-gnu/bin/ld: i386:x86-64 architecture of input file `/tmp/rustc.CNu40ABKRiEL/librustc_llvm-9ea7cfc2f58bb7a4.rlib(compatibility.o)' is incompatible with i386 output
[01:41:32]           /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../x86_64-unknown-linux-gnu/bin/ld: i386:x86-64 architecture of input file `/tmp/rustc.CNu40ABKRiEL/librustc_llvm-9ea7cfc2f58bb7a4.rlib(compatibility-debug_list.o)' is incompatible with i386 output
...
<snip>
...
[01:41:32]           /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../x86_64-unknown-linux-gnu/bin/ld: i386:x86-64 architecture of input file `/tmp/rustc.CNu40ABKRiEL/librustc_llvm-9ea7cfc2f58bb7a4.rlib(thread.o)' is incompatible with i386 output
[01:41:32]           /rustroot/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../x86_64-unknown-linux-gnu/bin/ld: i386:x86-64 architecture of input file `/tmp/rustc.CNu40ABKRiEL/librustc_llvm-9ea7cfc2f58bb7a4.rlib(regex.o)' is incompatible with i386 output
[01:41:32]           collect2: error: ld returned 1 exit status
[01:41:32]           
[01:41:32] 
[01:41:32] error: aborting due to previous error
[01:41:32] 
[01:41:33] error: Could not compile `rustc_trans`.
```

</details>
